File: lintcode/longest-substring-without-repeating-characters.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/longest-substring-without-repeating-characters/

class Solution:
    # @param s: a string
    # @return: an integer
    def lengthOfLongestSubstring(self, s):
        # write your code here
        if len(s) == 0:
            return 0

        st = set()
        st.add(s[0])
        istart = 0
        iend = 1
        res = 1
        while iend < len(s):
            if s[iend] in st:
                while s[istart] != s[iend]:
                    st.remove(s[istart])
                    istart += 1
                istart += 1
            else:
                st.add(s[iend])
                res = max(res, iend - istart + 1)
            iend += 1
        return res



if __name__ == '__main__':
    s = Solution()
    print s.lengthOfLongestSubstring('abcabcbb')
    print s.lengthOfLongestSubstring('bbbbb')

File: leetcode/longest-substring-without-repeating-characters.cc
// Question: https://leetcode.com/problems/longest-substring-without-repeating-characters/

#include <string>
#include <iostream>
#include <set>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.empty()) {
            return 0;
        }
        int ans = 1;
        set<char> hs;
        hs.insert(s.at(0));
        int j = 0;
        int i = 1;
        for (i = 1; i < s.size(); ++i) {
            if (hs.count(s.at(i)) > 0) {
                ans = max(ans, i-j);
                while (j < i) {
                    if (s.at(j) == s.at(i)) {
                        ++j;
                        break;
                    }
                    hs.erase(hs.find(s.at(j)));
                    ++j;
                }
            } else {
                hs.insert(s.at(i));
            }
        }
        return max(ans, i-j);
    }
};

int main(int argc, char** argv) {
    Solution s;
    cout << s.lengthOfLongestSubstring("c") << endl;
    cout << s.lengthOfLongestSubstring("cd") << endl;
    cout << s.lengthOfLongestSubstring("abccd") << endl;
    cout << s.lengthOfLongestSubstring("bbbbb") << endl;
}
